tests/test_shapely.py

```diff
@@ -0,0 +1,143 @@
+            [89588.192091767, 6894230.248628867],
+            [89574.61111389, 6894228.638802719]
+         ],
+         [
+            [1489610.344670435, 6894182.466199101],
+            [89625.985058891, 6894184.258949757],
+            [89629.547282597, 6894153.270030369],
+            [89613.918026089, 6894151.458993318],
+            [89610.344670435, 6894182.466199101]
+         ]
+      ]
+   ]
+}"""
+
+SAMPLE_POINT_GEOJSON = '{"type":"Point","coordinates":[1.2, 3.4]}'
+
+
+@pytest.fixture
+def shapely_conn(conn):
+    from psycopg.types.shapely import register_shapely
+
+    info = TypeInfo.fetch(conn, "geometry")
+    register_shapely(info, conn)
+    return conn
+
+
+def test_no_adapter(conn):
+    from shapely.geometry import Point
+
+    point = Point(1.2, 3.4)
+    with pytest.raises(ProgrammingError, match="cannot adapt type Point"):
+        conn.execute("SELECT pg_typeof(%s)", [point]).fetchone()[0]
+
+
+def test_with_adapter(shapely_conn):
+    from shapely.geometry import Point, Polygon
+
+    SAMPLE_POINT = Point(1.2, 3.4)
+    SAMPLE_POLYGON = Polygon([(0, 0), (1, 1), (1, 0)])
+
+    assert (
+        shapely_conn.execute(
+            "SELECT pg_typeof(%s)",
+            [SAMPLE_POINT],
+        ).fetchone()[0]
+        == "geometry"
+    )
+
+    assert (
+        shapely_conn.execute(
+            "SELECT pg_typeof(%s)",
+            [SAMPLE_POLYGON],
+        ).fetchone()[0]
+        == "geometry"
+    )
+
+
+@pytest.mark.parametrize("fmt_out", [Format.TEXT, Format.BINARY])
+def test_write_read_shape(shapely_conn, fmt_out):
+    from shapely.geometry import Point, Polygon
+
+    SAMPLE_POINT = Point(1.2, 3.4)
+    SAMPLE_POLYGON = Polygon([(0, 0), (1, 1), (1, 0)])
+
+    fmt_placeholder = "%b" if fmt_out == Format.BINARY else "%t"
+
+    with shapely_conn.cursor(binary=fmt_out) as cur:
+        cur.execute(
+            """
+        create table sample_geoms(
+            id     INTEGER PRIMARY KEY,
+            geom   geometry
+        )
+        """
+        )
+        cur.execute(
+            f"insert into sample_geoms(id, geom) VALUES(1, {fmt_placeholder})",
+            (SAMPLE_POINT,),
+        )
+        cur.execute(
+            f"insert into sample_geoms(id, geom) VALUES(2, {fmt_placeholder})",
+            (SAMPLE_POLYGON,),
+        )
+
+        cur.execute("select geom from sample_geoms where id=1")
+        result = cur.fetchone()[0]
+        assert result == SAMPLE_POINT
+
+        cur.execute("select geom from sample_geoms where id=2")
+        result = cur.fetchone()[0]
+        assert result == SAMPLE_POLYGON
+
+
+@pytest.mark.parametrize("fmt_out", [Format.TEXT, Format.BINARY])
+def test_match_geojson(shapely_conn, fmt_out):
+    from shapely.geometry import Point
+
+    SAMPLE_POINT = Point(1.2, 3.4)
+    with shapely_conn.cursor(binary=fmt_out) as cur:
+        cur.execute(
+            f"""
+            select ST_GeomFromGeoJSON('{SAMPLE_POINT_GEOJSON}')
```

Comment:
Maybe better to pass this as normal parameter instead of f-string? Same for others?

Reply:
Done